File: gfapy/line/edge/gfa2/to_gfa1.py
import gfapy

class ToGFA1:

  def _to_gfa1_a(self):
    """List of the field content of the line in GFA1.
    """
    at = self._alignment_type
    if at == "I":
      raise gfapy.RuntimeError(
        "Conversion of edge line from GFA2 to GFA1 failed\n"+
        "Edge represents an internal overlap:\n"+
        "Edge line: {}\n".format(str(self)))
    a = [ at ]
    if self._is_sid1_from():
      ol1 = self.get("sid1")
      ol2 = self.get("sid2")
    else:
      ol1 = self.get("sid2")
      ol2 = self.get("sid1")
    a.append(ol1.name)
    a.append(ol1.orient)
    a.append(ol2.name)
    a.append(ol2.orient)
    if self._alignment_type == "C":
      a.append(str(self.pos))
    try:
      self.overlap.validate(version = "gfa1")
    except:
      raise gfapy.RuntimeError(
        "Conversion of edge line from GFA2 to GFA1 failed\n"+
        "Overlap is invalid or not compatible with GFA1\n"+
        "Edge line: {}\n".format(str(self)))
    a.append(str(self.overlap))
    if not gfapy.is_placeholder(self.eid):
      a.append(gfapy.Field._to_gfa_tag(self.eid, "ID", datatype = "Z"))
    for fn in self.tagnames:
      a.append(self.field_to_s(fn, tag = True))
    return a

  @property
  def overlap(self):
    """Value of the GFA1 **overlap** field, if the edge is a link or containment.

    Returns:
      gfapy.Alignment.Placeholder or gfapy.Alignment.CIGAR

    Raises:
      gfapy.error.ValueError: If the edge is internal
    """
    self._check_not_internal("overlap")
    return self.alignment.complement() if self._is_sid1_from() else self.alignment

  @property
  def oriented_from(self):
    return self.sid1 if self._is_sid1_from() else self.sid2

  @property
  def oriented_to(self):
    return self.sid2 if self._is_sid1_from() else self.sid1

  @property
  def from_segment(self):
    """Value of the GFA1 **from_segment** field, if the edge is a link or containment.

    Returns:
      str or gfapy.line.segment.GFA2

    Raises:
      gfapy.error.ValueError: If the edge is internal.
    """
    return self.oriented_from.line

  @from_segment.setter
  def from_segment(self, value):
    """Set the field which will be returned by calling from_segment

    Parameters:
      value (str, gfapy.line.segment.GFA2)
    """
    self.oriented_from.line = value

  @property
  def from_orient(self):
    """Value of the GFA1 **from_orient** field.

    This method can only be applied if the edge is a link or containment.

    Returns:
      str: one of ["+", "-"]

    Raises:
      gfapy.error.ValueError: If the edge is internal.
    """
    return self.oriented_from.orient

  @from_orient.setter
  def from_orient(self, value):
    """Set the orientation of the field which will be returned by calling
    from_segment

    Parameters:
      value (str): one of ["+", "-"]
    """
    self.oriented_from.orient = value

  @property
  def to_segment(self):
    """Value of the GFA1 ``to_segment`` field.

    This method can only be applied if the edge is a link or containment.

    Returns:
      str or gfapy.line.segment.GFA2

    Raises:
      gfapy.error.ValueError: If the edge is internal.
    """
    return self.oriented_to.line

  @to_segment.setter
  def to_segment(self, value):
    """Set the field which will be returned by calling ``to_segment``.

    Parameters:
      value (str or gfapy.line.segment.GFA2)
    """
    self.oriented_to.line = value

  @property
  def to_orient(self):
    """Value of the GFA1 **to_orient** field.

    This method can only be applied if the edge is a link or containment.

    Returns:
      str : one of ["+", "-"]

    Raises:
      gfapy.error.ValueError: If the edge is internal.
    """
    return self.oriented_to.orient

  @to_orient.setter
  def to_orient(self, value):
    """Set the orientation of the field which will be returned by calling
    ``to_segment``.

    Parameters:
      value (str): one of ["+", "-"]
    """
    self.oriented_to.orient = value

  @property
  def pos(self):
    """Value of the GFA1 **pos** field, if the edge is a containment.

    Returns:
      int or gfapy.Lastpos

    Raises:
      gfapy.error.ValueError: If the edge is not a containment.
    """
    if self._alignment_type == "I":
      raise gfapy.ValueError("Line: {}\n".format(str(self)) +
                             "Internal alignment, pos is not defined")
    elif self._alignment_type == "L":
      raise gfapy.ValueError("Line: {}\n".format(str(self)) +
                             "Dovetail alignment, pos is not defined")
    elif self._alignment_type == "C":
      if gfapy.isfirstpos(self.beg1):
        return self.beg1 if (gfapy.isfirstpos(self.beg2) and
                             gfapy.islastpos(self.end2)) else self.beg2
      else:
        return self.beg1

  def _check_not_internal(self, fn):
    if self.is_internal():
      raise gfapy.ValueError(
        "Line: {}\n".format(str(self))+
        "Internal alignment, {} is not defined".format(fn))

  @staticmethod
  def _segment_role(begpos, endpos, orient):
    if gfapy.isfirstpos(begpos):
      if gfapy.islastpos(endpos):
        return "contained"
      elif orient == "+":
        return "pfx"
      else:
        return "sfx"
    else:
      if gfapy.islastpos(endpos):
        if orient == "+":
          return "sfx"
        else:
          return "pfx"
      else:
        return "other"

  def _check_GFA1_overlap_compatibility(self):
    pass

  def _is_sid1_from(self):
    sr1 = self._segment_role(self.beg1, self.end1, self.sid1.orient)
    sr2 = self._segment_role(self.beg2, self.end2, self.sid2.orient)
    if sr2 == "contained":
      return True
    elif sr1 == "contained":
      return False
    elif sr1 == "sfx" and sr2 == "pfx":
      return True
    elif sr2 == "sfx" and sr1 == "pfx":
      return False
    else:
      raise gfapy.ValueError(
        "Line: {}\n".format(str(self))+
        "Internal overlap, 'from' is undefined\n"+
        "Roles: segment1 is {} ({},{}), segment2 is {} ({},{})".format(sr1,
          self.beg1, self.end1, sr2, self.beg2, self.end2))

File: lib/rgfa/line/edge/gfa2/to_gfa1.rb
module RGFA::Line::Edge::GFA2::ToGFA1

  # @return [Array<String>] an array of fields of the equivalent line
  #   in GFA1, if the edge is a link or containment
  # @raise [RGFA::ValueError] if the edge is internal
  def to_gfa1_a
    at = alignment_type()
    if at == :internal
      raise RGFA::ValueError,
        "Internal overlap, cannot convert to GFA1\n#{self}"
    end
    a = [at]
    (sid1_from? ? [:sid1, :sid2] : [:sid2, :sid1]).each do |sid|
      ol = get(sid)
      a << ol.name.to_s
      a << ol.orient.to_s
    end
    if at == :C
      a << pos.to_s
    end
    a << overlap.to_s
    if !eid.placeholder?
      a << eid.to_gfa_tag(:id, datatype: :Z)
    end
    tagnames.each {|fn| a << field_to_s(fn, tag: true)}
    return a
  end

  # @return [RGFA::Alignment::Placeholder, RGFA::Alignment::CIGAR]
  #   value of the GFA1 +overlap+ field,
  #   if the edge is a link or containment
  # @raise [RGFA::ValueError] if the edge is internal
  def overlap
    check_not_internal(:overlap)
    sid1_from? ? alignment : alignment.complement
  end

  # @return [RGFA::OrientedLine] value of the GFA1 +from+ and +from_orient+
  #   fields, if the edge is a link or containment
  # @raise [RGFA::ValueError] if the edge is internal
  def oriented_from
    sid1_from? ? sid1 : sid2
  end

  # @return [RGFA::OrientedLine] value of the GFA1 +to+ and +to_orient+
  #   fields, if the edge is a link or containment
  # @raise [RGFA::ValueError] if the edge is internal
  def oriented_to
    sid1_from? ? sid2 : sid1
  end

  # @return [Symbol, RGFA::Line::Segment::GFA2] value of the GFA1 +from+ field,
  #   if the edge is a link or containment
  # @raise [RGFA::ValueError] if the edge is internal
  def from
    oriented_from.line
  end

  # Set the line of the field which will be returned by calling from
  # @param value [Symbol, RGFA::Line::Segment::GFA2]
  # @return [nil]
  def from=(value)
    oriented_from.line = value
  end

  # @return [:+, :-] value of the GFA1 +from_orient+ field,
  #   if the edge is a link or containment
  # @raise [RGFA::ValueError] if the edge is internal
  def from_orient
    oriented_from.orient
  end

  # Set the orientation of the field which will be returned by calling from
  # @param value [:+,:-]
  # @return [nil]
  def from_orient=(value)
    oriented_from.orient = value
  end

  # @return [Symbol, RGFA::Line::Segment::GFA2] value of the GFA1 +to+ field,
  #   if the edge is a link or containment
  # @raise [RGFA::ValueError] if the edge is internal
  def to
    oriented_to.line
  end

  # Set the line of the field which will be returned by calling to
  # @param value [Symbol, RGFA::Line::Segment::GFA2]
  # @return [nil]
  def to=(value)
    oriented_to.line = value
  end

  # @return [:+, :-] value of the GFA1 +to_orient+ field,
  #   if the edge is a link or containment
  # @raise [RGFA::ValueError] if the edge is internal
  def to_orient
    oriented_to.orient
  end

  # Set the orientation of the field which will be returned by calling to
  # @param value [:+,:-]
  # @return [nil]
  def to_orient=(value)
    oriented_to.orient = value
  end

  # @return [Integer] value of the GFA1 +pos+ field,
  #   if the edge is a containment
  # @raise [RGFA::ValueError] if the edge is not
  #   a containment
  def pos
    case alignment_type
    when :I
      raise RGFA::ValueError, "Line: #{self.to_s}\n"+
        "Internal alignment, pos is not defined"
    when :L
      raise RGFA::ValueError, "Line: #{self.to_s}\n"+
        "Dovetail alignment, pos is not defined"
    when :C
      if beg1.first?
        (beg2.first? and end2.last?) ? beg1 : beg2
      else
        beg1
      end
    end
  end

  private

  def check_not_internal(fn)
    if internal?
      raise RGFA::ValueError,
        "Line: #{self.to_s}\n"+
        "Internal alignment, #{fn} is not defined"
    end
  end

  # Role of a segment in an overlap, given coordinates and orientation.
  # @returns [Symbol] :pfx, :sfx, :contained, :other
  def segment_role(begpos, endpos, orient)
    if begpos.first?
      if endpos.last?
        return :contained
      else
        return orient == :+ ? :pfx : :sfx
      end
    else
      if endpos.last?
        return orient == :+ ? :sfx : :pfx
      else
        return :other
      end
    end
  end

  # @return [Boolean] does the sid1 correspond to from in GFA1?
  def sid1_from?
    sr1 = segment_role(beg1, end1, sid1.orient)
    sr2 = segment_role(beg2, end2, sid2.orient)
    if sr2 == :contained
      return true
    elsif sr1 == :contained
      return false
    elsif sr1 == :sfx and sr2 == :pfx
      return true
    elsif sr2 == :sfx and sr1 == :pfx
      return false
    else
      raise RGFA::ValueError, "Internal overlap, from undefined\n#{self}\n"+
         "Roles: segment1 is #{sr1}; segment2 is #{sr2}"
    end
  end

end
